### DISCUSSION TITLE
piece checking fails on 32-bit machines

### DISCUSSION THREAD

**Discussion by stefantalpalaru** [transmission/transmission#4542] [Support] [answered]
If I'm reading this correctly, pieces are checked only when seeding them: https://github.com/transmission/transmission/blob/cc4cbff0497564b320cc5f9062c9b756220c37b4/libtransmission/peer-msgs.cc#L2078

If this check fails, our bad piece is not re-downloaded, but the whole torrent is stopped and an error flag is set, asking the user to manually start a verification process.

This creates all those issues mentioned here: https://github.com/transmission/transmission/pull/4178

How about changing the logic so every piece is checked immediately after download and is marked as missing if the check fails, so it can be downloaded again?

**Comment by ckerr** [comment]
Transmission does check every piece during download when it becomes complete. That's what makes 4178 so vexing :smile: :crying_cat_face:

**Comment by stefantalpalaru** [comment]
> Transmission does check every piece during download

Where? It can't be using `ensurePieceIsChecked()` for that, because I grepped for it and checked every occurrence.

**Comment by stefantalpalaru** [comment]
The check during download is done here, in `tr_torrentGotBlock()`: https://github.com/transmission/transmission/blob/cc4cbff0497564b320cc5f9062c9b756220c37b4/libtransmission/torrent.cc#L2218

I put a debug print in `tr_torrent::checkPiece()` and all pieces are indeed checked during download. The only significant difference from `tr_torrent::ensurePieceIsChecked()` is the lack of `checked_pieces_.set(piece, checked);` which leads to this redundant check during seeding.

The question remains: how can this second piece check fail?

**Reply by ckerr** [reply]
Great question. I wish I knew.

Just brainstorming here.. one thing we could try, maybe is to test again after flushing a piece out to disk. Right now when we test the newly-completed piece it's partially (or fully) in the [write cache](https://github.com/transmission/transmission/blob/main/libtransmission/cache.h). Then we write it to disk _after_ the test passes.

If there was a bug in the writing, or if there was a bug in flushing from the write cache, or if there was an unhandled write errno somewhere, then forcing a piece test _from disk_ could help surface the bug...?

**Reply by ckerr** [reply]
Also... this could be a red herring, but FWIW I am just not seeing this issue _at all_.

As in, I seed ... a lot ... of torrents and have gotten the dreaded "piece is corrupt!" message maybe twice in the last year.

I'm on ext4 on Linux and I don't use the `incomplete-dir` feature, so if this is a Transmission bug...

- _maybe_ it's somewhere in `#ifdef` code for other filesystems
- _maybe_ it's somewhere in `#ifdef` code for not-Linux
- _maybe_ it's in the code that moves content from the incomplete dir to the complete dir

Off the top of my head I don't have any likely suspects. That part of the code has CI tests. ... but still, just brainstorming here about where a bug _could_ be hiding.

**Reply by stefantalpalaru** [reply]
I tried to use the Address Sanitizer to hunt for memory corruption, but only found an unrelated use-after-free in libcurl: https://github.com/curl/curl/issues/10273

An interesting, but confusing data point is that piece checking is broken much more often on 32-bit ARM. I tested the latest 4.0.0 beta there and manually verifying the local data invalidated as much as half of the data on disk. Went back to 3.00 and the same pieces all passed validation. There's something wrong with the verification process, but I read the code and can't put my finger on anything suspicious.

There's nothing wrong with the disk - SATA SSD, ext4 - on my little ARM SoC. There's also no disk issue on my desktop - ZFS with on-disk checksumming that shows no errors on scrub.

**Reply by ckerr** [reply]
> Went back to 3.00 and the same pieces all passed validation.

When you say "same pieces" you mean you downloaded once and pointed both 3.00 and 4.0.0 beta at that same one download, rather than downloaded in 4 and then re-downloaded in 3? I'm pretty sure the answer is "yes" but want to make sure I'm reading the clues correctly.

Given that idea, other things to try

- What crypto backends are your 3.00 and 4.0.0 beta built with?
- If you took that same downloaded data that 3.00 ARM liked but 4.0.0 beta ARM didn't, what happens if you verify with 4.0.0 beta not-ARM?
- If you took that same downloaded data that 3.00 ARM lilked but 4.0.0 beta ARM didn't, what happens if you verify with 4.0.0 beta ARM, but built with a different crypto backend?

**Reply by stefantalpalaru** [reply]
> you downloaded once and pointed both 3.00 and 4.0.0 beta at that same one download

Yes. Torrents completed and being seeded before the Transmission upgrade.

> What crypto backends are your 3.00 and 4.0.0 beta built with?

Same OpenSSL on Gentoo. All software compiled from source.

I'll try the data move test.

**Reply by stefantalpalaru** [reply]
Torrent completed and seeding on ARM, with Transmission daemon 3.00. Triggered a verification and all is good. Copied it to an AMD64 machine.

Upgraded the ARM daemon to Transmission HEAD, triggered a verification and an error appeared, leaving the torrent at 78% completion. The error: "Invalid argument (/full/path/to/file.mkv)". Two more local data verifications stop at the same point, with the same error. Restarted the daemon, triggered another verification through "Transmission Remote" which stops at the 78%, but without an error message. Another daemon restart and the error came back, associated with a log message from here: https://github.com/transmission/transmission/blob/770da79cf39e41772bc46dfe61f671cd7a1f5de8/libtransmission/inout.cc#L177

Ran md5sum on both the ARM original and the AMD64 copy and they remained identical.

On AMD64, I started a GTK instance of Transmission HEAD from a Magnet link. It discovered it had 100% of data available (the one copied over from the ARM system) and started seeding.

Weird stuff: on ARM, two other paused torrents show this error with HEAD: «No data found! Ensure your drives are connected or use "Set Location". To re-download, remove the torrent and re-add it.». Going back to 3.00 makes all errors disappear.

**Reply by ckerr** [reply]
OK, so if you downloaded once (using 3.00 on ARM) and that local data verifies 100% on 3.00 on ARM _and_ in head on AMD64, but only  to 78% in head on ARM...

What is the size, in bytes, of `file.mkv`? And what is `sizeof(size_t)` on your ARM box? ... In particular I'm wondering if there's some code in `main` should be using `uint64_t` but is instead using `size_t` that caps out as a uint32 on ARM.

**Reply by stefantalpalaru** [reply]
There are multiple files in the torrent. Here's some size info:

```text
  Piece Count: 3552
  Piece Size: 4.00 MiB
  Total Size: 14.90 GB
```

**Reply by stefantalpalaru** [reply]
> And what is sizeof(size_t) on your ARM box?

4

**Reply by ckerr** [reply]
@stefantalpalaru is the `/full/path/to/file.mkv` mentioned in https://github.com/transmission/transmission/discussions/4542#discussioncomment-4723658 larger than 4 GiB or  4.29 GB?

**Reply by stefantalpalaru** [reply]
No. The MKV file in the middle of which verification stops is only 38.63 MB.

**Reply by ckerr** [reply]
Drat.

What about if you use `transmission-show --unsorted /path/to/that/torrent.torrent` and add together the sizes of all the files **before and including** `file.mkv`, is _that_ larger than 4 GiB or 4.29 GB?

**Reply by stefantalpalaru** [reply]
It's 11.37 GiB.

**Reply by ckerr** [reply]
I'm able to reproduce this in a 32bit environment.

**Comment by GaryElshaw** [comment]
I'd be overjoyed to test any pr's :-)

**Reply by ckerr** [reply]
@GaryElshaw are you experiencing "Piece #X is corrupt!" issues so frequently that you can test this on the regular?

**Reply by GaryElshaw** [reply]
For me, it's more that i'm regularly seeing a lot of corruption (5 - 10%), and rarely seeing "Piece #X is corrupt!". I'd be really happy to help solve this mystery.

**Reply by ckerr** [reply]
Do you see the same corruption level on both public and private trackers?

**Reply by GaryElshaw** [reply]
I exclusively use public trackers.

**Reply by ckerr** [reply]
My guess is that you're experiencing impostor peers out there that are intentionally poisoning the swarm. Detecting sus peers is a valid issue, but it's not the same as this checksums-fail-when-reading-from-disk one.

_If_ that is what's going on then it will probably persist :disappointed:  until Transmission supports BitTorrent v2. Merkle trees make bad peer detection much easier.

**Reply by GaryElshaw** [reply]
There's this weird degree of unpredictability. I will frequently begin downloading, wait until around 30% (or when the peers appear to be ebbing in numbers or the download rate drops to double-digits), verify, find out how much data is corrupt, and then often use another client to download the remaining data.

**Comment by ckerr** [comment]
I think #4007 is the same as the issue we're discussing here.

**Reply by GaryElshaw** [reply]
Sorry, i should have been clearer, i'm seeing this on MacOS.

**Reply by ckerr** [reply]
Point taken. My _guess_ is the UI flavor won't affect this issue, but who knows.

**Comment by ckerr** [comment, marked as answer]
Thaks to Ștefan's clues :green_heart: I've been able to reproduce the issue locally and have converted this into issue #4626.
